Add Semantic Scholar translator

GuyAglionby opened the pull request:
Semantic scholar is currently served by the embedded metadata translator, but this doesn't capture everything so I went ahead and wrote a translator.

In terms of PDF downloading, it'll work for most articles where a link is available. However, there is a corner case where the first link on the "download button" is not to a .pdf but to another website, and the direct .pdf link is available under a dropdown. An example of this is [here](https://www.semanticscholar.org/paper/Foundations-of-Statistical-Natural-Language-Proces-Manning-Sch%C3%BCtze/06fd7d924d499fbc62ccbcc2e458fb6c187bcf6f). The issue is that the HTML for the direct link is not present until the dropdown button is pressed, and I'm not sure how to simulate a click on it. `onclick` isn't bound to anything, I assume as a consequence of using React. The rest of the code is there if you have any insight into how to do that; I'm not too sure.

Let me know what you think. Thanks!

adam3smith commented:
@GuyAglionby I don't think we can actually simulate on-click, but we can look at traffic and figure out what the internal API calls are and use those. In this case, though, I think there's a simpler solution. If you look for the PDF url (minus the session specific _?ga.... bit), you'll find it in the site JSON under links. You should be able to grab&use that.

zuphilip commented:
( @adam3smith Simulating click event may also be possible, see for example https://github.com/zotero/translators/blob/master/ScienceDirect.js#L81-L103 )

gdlg commented:
We can get the PDF url from the site DATA variable which is defined in a script on the page using something like:

```javascript
var script = ZU.xpath(doc, '//script');

for (var i = 0; i < script.length; i++) {
    if (script[i].text.startsWith("var DATA")) {
        eval(script[i].text);
        links = DATA[0].resultData.paper.links;
        for (var j = 0; j < links.length; j++) {
            if (links[j].linkType == "s2") {
                pdf = links[j].url;
            }
        }
        break;
    }
}
```

adam3smith commented:
@gdlg yes, that's what I was referring to above, thanks for putting the code in.

GuyAglionby commented:
All sounds good thanks! Will send a PR tomorrow with it all
________________________________
From: Sebastian Karcher <notifications@github.com>
Sent: Friday, December 15, 2017 8:34:48 PM
To: zotero/translators
Cc: Guy Aglionby; Mention
Subject: Re: [zotero/translators] Add Semantic Scholar translator (#1497)


@gdlg<https://github.com/gdlg> yes, that's what I was referring to above, thanks for putting the code in.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/zotero/translators/pull/1497#issuecomment-352105297>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABRnqwEdpXlOVPCvSJmcFXZ1hD3h0fgsks5tAtfogaJpZM4RC8Iq>.

GuyAglionby commented:
Ok, have implemented that way of grabbing the PDF and added a test for it, thanks! In the course of doing this I found that the BibTeX they produce has an unbalanced bracket error for {e,i,o,u} with umlauts -- for example in the title [here](https://www.semanticscholar.org/paper/Interleukin-7-mediates-the-homeostasis-of-na%C3%AFve-an-Schluns-Kieper/aee7b854bed51120fe356a5792dfb22fec7cf2ae). I've implemented a fix for this and added some tests as well.

Let me know what you think! Cheers.

adam3smith commented:
Sorry it's taken a bit to review -- is there a reason you're not trying to import multiples from search results, e.g. https://www.semanticscholar.org/search?q=test&sort=relevance that would seem to be super useful

GuyAglionby commented:
Not to worry. There was no particular reason for missing those out -- just wasn't part of my use case. I've added support for the search and author pages :)

zuphilip commented on Semantic Scholar.js:
```diff
@@ -0,0 +1,526 @@
+function createUmlautReplacement(char) {
+	// There's this issue where e,i,o,u characters with umlauts have unbalanced
+	// braces in the Semantic Scholar BibTeX, which kills the Zotero translator.
+	return [new RegExp('{\\\\"{' + char + '}[^}]', 'g'), '{\\"' + char + '}'];
```
Anyway, it looks unneeded to have two functions for fixing the bibtex (see also comment below).

GuyAglionby replied:
I've double checked and I don't think this is the case -- the tests confirm that the correct spellings are extracted. Let me know if I'm missing something obvious though

zuphilip commented on Semantic Scholar.js:
Dito

zuphilip commented on Semantic Scholar.js:
This looks like volume and issue number.

GuyAglionby commented on Semantic Scholar.js:
```diff
@@ -0,0 +1,526 @@
+	umlautReplacements.forEach(function(replacement) {
+		bibtex = bibtex.replace(replacement[0], replacement[1]);
+	});
+	return bibtex;
```
I agree it's better as a single function, thanks for the tip. I've only found problems on the site with `[eEiIoOuU]` with umlauts, but as the regex doesn't break properly formatted characters I'll change it to `[a-zA-z]` :)

GuyAglionby commented on Semantic Scholar.js:
From looking at a few examples, you're correct in saying this is how they format the issue and volume -- I've amended the code so it takes care of this.

In this case, I think this is actually the issue number, given [Nature's current issue](https://www.nature.com/nature/current_issue.html) is 7685, and I've seen this particular article cited as issue 7000 in a few papers.

GuyAglionby commented:
Thanks for your feedback! I've made the changes you mention; I agree with the majority of them and have left some comments where I was maybe unclear.

All tests of course updated & passing.

Cheers :)

adam3smith commented:
Great, thanks! (I forgot to update the tests after changing the label for the PDF attachments, so tests will come up as mismatch -- we can fix later)

GuyAglionby commented:
Fab, cheers :)

zuphilip commented:
Thank you @GuyAglionby for the work here!
